### PR TITLE
expression: support datetime type for user variable

### PR DESCRIPTION
**Pull request by lzmhhh123**
Signed-off-by: lzmhhh123 <lzmhhh123@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25248

### What is changed and how it works?

What's Changed: support datetime type for user variable

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- none

### CODE REVIEW

File: expression/builtin_other.go [hunk 1]
```diff
@@ -844,6 +848,34 @@ func (b *builtinSetIntVarSig) evalInt(row chunk.Row) (int64, bool, error) {
...
+
+func (b *builtinSetTimeVarSig) evalTime(row chunk.Row) (types.Time, bool, error) {
+	sessionVars := b.ctx.GetSessionVars()
+	varName, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return types.ZeroTime, isNull, err
+	}
+	datum, err := b.args[1].Eval(row)
+	if err != nil || datum.IsNull() {
+		return types.ZeroTime, isNull, handleInvalidTimeError(b.ctx, err)
```
Comment (wshwsh12) [inline expression/builtin_other.go:869]:
```suggestion
		return types.ZeroTime, datum.IsNull(), handleInvalidTimeError(b.ctx, err)
```

File: expression/builtin_other_test.go [hunk 2]
```diff
@@ -193,10 +196,16 @@ func (s *testEvaluatorSuite) TestGetVar(c *C) {
 		{"e", int64(3)},
 		{"f", float64(2.5)},
 		{"g", dec},
+		{"h", timeDec},
 	}
 	for _, kv := range sessionVars {
 		s.ctx.GetSessionVars().Users[kv.key] = types.NewDatum(kv.val)
-		tp := types.NewFieldType(mysql.TypeVarString)
+		var tp *types.FieldType
+		if _, ok := kv.val.(*types.MyDecimal); ok {
```
Comment (wshwsh12) [inline expression/builtin_other_test.go:204]: Why check it .. And why the type is *types.MyDeciaml?

**Review by wshwsh12 — COMMENTED** [review]
Please fix ci. Rest LGTM

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**